### PR TITLE
build: remove setup.cfg, in favour of pyproject.toml

### PR DESCRIPTION
**Pull request by bdura**
This PR proposes to remove `setup.cfg` in favour of pyproject.toml.

### CODE REVIEW

File: .github/workflows/tests.yml [hunk 1]
```diff
@@ -28,25 +28,17 @@ jobs:
...
-
-      - uses: actions/cache@v3
-        with:
-          path: ~/.cache/pre-commit
-          key: pre-commit|${{ env.PY }}|${{ hashFiles('.pre-commit-config.yaml') }}
-
       - name: Install pre-commit
         run: |
-          pip install 'pre-commit>=3.2.0,<4.0.0'
+          pip install 'pre-commit==3.3.2'
```
Comment (rmitsch) [inline .github/workflows/tests.yml:33]: Why the exact pin now?

File: pyproject.toml [hunk 2]
```diff
@@ -1,7 +1,55 @@
+[project]
+name = "weasel"
+version = "0.0.0"
+authors = [
+    {name = "Explosion", email = "contact@explosion.ai"},
+]
+description = "Weasel: A small and easy workflow system"
+readme = "README.md"
+requires-python = ">=3.8"
```
Comment (rmitsch) [inline pyproject.toml:9]: Unrelated to this PR, but isn't this a problem since we support Python 3.6+ in spaCy?

File: pyproject.toml [hunk 3]
```diff
@@ -1,7 +1,55 @@
...
+
+[tools.setuptools.dynamic]
+version = {attr = "weasel.__version__"}
+
+[project.entry-points."console_scripts"]
+weasel = "weasel:app"
+
 [build-system]
-requires = ["setuptools"]
+requires = ["setuptools", "setuptools-scm"]
```
Comment (rmitsch) [inline pyproject.toml:45]: Why is this necessary?

**Review by rmitsch — COMMENTED** [review]
What's the motivation for switching to `pyproject.toml`? (Also this is in unassigned, but still marked as draft. Not sure if you want reviews for this already)

File: pyproject.toml (hunk 2 above)
Comment (bdura) [reply pyproject.toml:9]: I might be mistaken, but I think spaCy v4 will drop support for Python 3.6 and 3.7 since they've reached end-of-life (Python 3.7 will reach EOL on the 27th).

It makes things much easier for us, we no longer have to rely on brittle backports.

File: pyproject.toml (hunk 3 above)
Comment (bdura) [reply pyproject.toml:45]: It's not, thanks. It might be useful in something like `spacy-llm` though, it would remove the need for a MANIFEST file for template inclusion.

File: .github/workflows/tests.yml (hunk 1 above)
Comment (bdura) [reply .github/workflows/tests.yml:33]: Well, I figured we might as well use stricter constraints to make sure everything is reproducible... Although pinning the patch version is probably not a great idea.

**Comment by bdura** [comment]
> What's the motivation for switching to pyproject.toml? (Also this is in unassigned, but still marked as draft. Not sure if you want reviews for this already)

Thanks for the review @rmitsch! I created this PR to see what a `setup.cfg/py`-free Weasel would look like, but it's not consistent with other Explosion projects so I don't actually expect it to be merged. Feel free to close it if the experiment doesn't pan out.

File: .github/workflows/tests.yml (hunk 1 above)
Comment (rmitsch) [reply .github/workflows/tests.yml:33]: I think usually we pin to the minor at most, as to avoid dependency conflicts for users in their venvs (and it's rare that there are breaking changes in between minors).

File: pyproject.toml (hunk 2 above)
Comment (rmitsch) [reply pyproject.toml:9]: Ah, ok. In that case this should target the v4 branch.

**Comment by rmitsch** [comment]
>  I created this PR to see what a setup.cfg/py-free Weasel would look like, but it's not consistent with other Explosion projects so I don't actually expect it to be merged. Feel free to close it if the experiment doesn't pan out.

Ah, ok. My guess is that we prefer consistency here, but curious about what @svlandeg and @adrianeboyd think.

File: .github/workflows/tests.yml (hunk 1 above)
Comment (svlandeg) [reply .github/workflows/tests.yml:33]: This is just for the tests though right, so it's not that big a deal?

File: pyproject.toml (hunk 2 above)
Comment (svlandeg) [reply pyproject.toml:9]: I don't think we want to make this change as part of this PR right now - for now we should just keep supporting 3.6 and 3.7 - the backports are all already there. Once we do decide to drop support from spaCy, it won't be difficult to just sweep across and clean up.

**Comment by adrianeboyd** [comment]
At this point I do think I'd rather keep using `setup.cfg` across the board unless we're at the point where we want to migrate all projects.

I'm not sure about the side effects with projects that use cython in particular and I think we've run into issues with pytest settings and a few other things being slightly different in practice, although again, mainly it would just need more time for testing in a wider range of projects.

**Comment by bdura** [comment]
I'll close this PR for now, feel free to revive it!
